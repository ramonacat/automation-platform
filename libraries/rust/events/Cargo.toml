[package]
name = "events"
version = "0.1.0"
edition = "2021"
license = "BSD-3-Clause"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1.0.154", features = ["derive"] }
serde_json = "1.0.94"
uuid = { version = "1.3.0", features=["v4", "serde"] }
tokio = { version = "1.26.0", features = ["full"] }
tracing = "0.1.37"
async-trait = "0.1.66"
thiserror = "1.0.39"
rpc-support = { path="../rpc-support"}
platform = { path="../platform"}
futures = "0.3.26"

[build-dependencies]
message-compiler={path= "../../../tools/message-compiler" }