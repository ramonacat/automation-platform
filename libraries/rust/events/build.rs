use message_compiler::type_checking::TypeChecker;
use std::error::Error;
use std::fs::read_to_string;

fn main() -> Result<(), Box<dyn Error>> {
    println!("cargo:rerun-if-changed=events.evd");

    let parser = message_compiler::parsing::grammar::RFileParser::new();
    let events = read_to_string("events.evd")?;
    let ast = parser.parse(&events).unwrap();

    let type_checker = TypeChecker::new();
    let typed_file = type_checker.check(&ast)?;

    let rust = message_compiler::compiler_rust::compile(typed_file);
    let out_dir = std::env::var_os("OUT_DIR").unwrap();
    let dest_path = std::path::Path::new(&out_dir).join("structs.rs");
    std::fs::write(dest_path, rust).unwrap();

    Ok(())
}
