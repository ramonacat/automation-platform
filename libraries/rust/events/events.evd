metadata {
    source: string,
    correlation_id: guid,
}

struct FileOnMountPath {
    path: string,
    mount_id: string
}

struct Event {
    id: guid,
    created_time: instant,
    data: EventKind
}

struct SubscribeRequest {
    id: guid,
    from: instant?,
}

enum EventKind {
    FileCreated(path: FileOnMountPath),
    FileDeleted(path: FileOnMountPath),
    FileChanged(path: FileOnMountPath),
    FileMoved(from: FileOnMountPath, to: FileOnMountPath),
}

rpc {
    send_event(Event) -> void;
    subscribe(SubscribeRequest) -> stream Event;
}
