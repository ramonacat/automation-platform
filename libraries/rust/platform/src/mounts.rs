use std::collections::HashMap;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Debug, Error, PartialEq, Eq)]
pub enum MountError {
    #[error("No mounts were defined")]
    NoMountsDefined,
    #[error("The given absolute path is not within the mount")]
    UnableToMakeRelative,
    #[error("Failed to make the path mount-relative")]
    PathNotInMount,
    #[error("Mount with id \"{0}\" does not exist")]
    NoSuchMount(String),
}

pub struct Provider {
    mounts: HashMap<String, Mount>,
}

#[derive(Debug, Clone)]
pub struct Mount {
    id: String,
    path: PathBuf,
}

impl Mount {
    #[must_use]
    pub const fn new(id: String, path: PathBuf) -> Self {
        Self { id, path }
    }

    #[must_use]
    pub fn id(&self) -> &str {
        &self.id
    }

    #[must_use]
    pub fn path(&self) -> &Path {
        &self.path
    }
}

#[derive(Debug, Clone)]
pub struct PathInside {
    // FIXME should the mount be a reference?
    mount_id: String,
    path: String,
}

impl PathInside {
    #[must_use]
    pub const fn new(mount_id: String, path: String) -> Self {
        Self { mount_id, path }
    }

    #[must_use]
    pub fn mount_id(&self) -> &str {
        &self.mount_id
    }

    #[must_use]
    pub fn path(&self) -> &str {
        &self.path
    }
}

impl Provider {
    #[must_use]
    pub fn new(mounts: Vec<Mount>) -> Self {
        Self {
            mounts: mounts
                .into_iter()
                .map(|mount| (mount.id.clone(), mount))
                .collect(),
        }
    }

    #[must_use]
    pub fn from_raw_string(directories_from_env: &str) -> Self {
        Self::new(
            directories_from_env
                .split(',')
                .map(|x| x.split(':').collect())
                .map(|x: Vec<&str>| Mount::new(x[1].into(), x[0].into()))
                .collect(),
        )
    }

    // FIXME can we avoid the clones?
    #[must_use]
    pub fn mounts(&self) -> Vec<Mount> {
        self.mounts.values().cloned().collect()
    }

    // FIXME this needs some careful security review
    /// # Errors
    /// If the path is not within any of the mounts, or if no mounts are defined
    pub fn path_inside_from_filesystem_path(&self, path: &Path) -> Result<PathInside, MountError> {
        let mut result = Err(MountError::NoMountsDefined);
        for mount in self.mounts.values() {
            result = self.path_inside_from_filesystem_path_with_mount(path, mount);

            if result.is_ok() {
                return result;
            }
        }

        result
    }

    /// # Errors
    /// - `MountError::UnableToMakeRelative` if the pathdiff fails
    /// - `MountError::PathNotInMount` if the path is not within the mount
    pub fn path_inside_from_filesystem_path_with_mount(
        &self,
        path: &Path,
        mount: &Mount,
    ) -> Result<PathInside, MountError> {
        pathdiff::diff_paths(path, &mount.path).map_or(
            Err(MountError::UnableToMakeRelative),
            |relative_path| {
                if relative_path.starts_with("..") {
                    Err(MountError::PathNotInMount)
                } else {
                    Ok(PathInside {
                        mount_id: mount.id.clone(),
                        path: relative_path.as_path().to_string_lossy().replace('\\', "/"),
                    })
                }
            },
        )
    }

    /// # Errors
    /// Returns an error if there's no mount for the given path
    pub fn mount_relative_to_filesystem_path(
        &self,
        path_inside: PathInside,
    ) -> Result<PathBuf, MountError> {
        let mount = self
            .mounts
            .get(&path_inside.mount_id)
            .ok_or_else(|| MountError::NoSuchMount(path_inside.mount_id.clone()))?;
        let mut path = mount.path.clone();
        path.push(path_inside.path);

        Ok(path)
    }

    /// # Errors
    /// Returns `Err` if the mount is not found.
    pub fn mount_relative_to_filesystem_path_by_mount_id(
        &self,
        mount_id: &str,
        path_inside: &str,
    ) -> Result<PathBuf, MountError> {
        let mount = self
            .mounts
            .get(mount_id)
            .ok_or_else(|| MountError::NoSuchMount(mount_id.into()))?;
        let mut path = mount.path.clone();
        path.push(path_inside);

        Ok(path)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    pub fn can_create_mount_relative_path_from_absolute() {
        let mount = Mount::new("some_id".into(), Path::new("/tmp/a/").to_path_buf());
        let provider = Provider::new(vec![mount.clone()]);

        let result = provider
            .path_inside_from_filesystem_path_with_mount(Path::new("/tmp/a/b/c/"), &mount)
            .unwrap();

        assert_eq!(result.path(), "b/c");
    }

    #[test]
    pub fn can_find_the_matching_mount_from_absolute_path() {
        let mount_a = Mount::new("mount_a".into(), Path::new("/tmp/a/").to_path_buf());
        let mount_b = Mount::new("mount_b".into(), Path::new("/tmp/b/").to_path_buf());
        let provider = Provider::new(vec![mount_a, mount_b]);

        let result = provider
            .path_inside_from_filesystem_path(&Path::new("/tmp/b/c"))
            .unwrap();

        assert_eq!(result.mount_id(), "mount_b");
    }

    #[test]
    pub fn will_error_if_path_is_in_none_of_the_mounts() {
        let mount_a = Mount::new("mount_a".into(), Path::new("/tmp/a/").to_path_buf());
        let mount_b = Mount::new("mount_b".into(), Path::new("/tmp/b/").to_path_buf());

        let provider = Provider::new(vec![mount_a, mount_b]);
        let result = provider.path_inside_from_filesystem_path(Path::new("/tmp/c/d"));

        assert_eq!(result.unwrap_err(), MountError::PathNotInMount);
    }

    #[test]
    pub fn mount_has_an_id() {
        let mount = Mount::new("some_id".into(), Path::new("/tmp/a/").to_path_buf());

        assert_eq!(mount.id(), "some_id");
    }
}
