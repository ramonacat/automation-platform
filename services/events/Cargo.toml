[package]
name = "svc-events"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tokio = { version = "1.20.1", features = ["full"] }
tracing="0.1.36"
tracing-subscriber = "0.3.15"
serde = { version = "1.0.143", features=["derive"] }
serde_json="1.0.74"
tokio-postgres = { version = "0.7.5", features=["with-uuid-1", "with-time-0_3", "with-serde_json-1"] }
uuid = { version = "1.1.2", features=["v4"] }
postgres-native-tls="0.5.0"
native-tls="0.2.10"
thiserror = "1.0.30"
time = { version = "0.3.11", features=["local-offset", "parsing", "serde", "formatting"] }
rpc-support = { path="../../libraries/rust/rpc-support"}
async-trait = "0.1.57"

platform = { path = "../../libraries/rust/platform" }
events = { path = "../../libraries/rust/events" }
