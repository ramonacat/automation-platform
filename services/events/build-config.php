<?php

use Ramona\AutomationPlatformLibBuild\Actions\Docker\BuildNixifiedDockerImage;
use Ramona\AutomationPlatformLibBuild\Actions\Kubernetes\GenerateKustomizeOverride;
use Ramona\AutomationPlatformLibBuild\Actions\Kubernetes\KustomizeApply;
use Ramona\AutomationPlatformLibBuild\Actions\Kubernetes\KustomizeOverride;
use Ramona\AutomationPlatformLibBuild\Context;
use Ramona\AutomationPlatformLibBuild\Definition\BuildDefinitionBuilder;
use Ramona\AutomationPlatformLibBuild\Targets\DefaultTargetKind;
use Ramona\AutomationPlatformLibBuild\Targets\TargetId;


return static function (BuildDefinitionBuilder $builder) {
    $builder->addRustTargetGenerator();

    $builder->addTarget(
        'image-service-docker-build',
        new BuildNixifiedDockerImage('image-service', 'svc-events')
    );

    $builder->addTarget(
        'image-migrations-docker-build',
        new BuildNixifiedDockerImage('image-migrations', 'svc-events-migrations', nixFilePath: 'docker/migrations.nix')
    );

    $builder->addTarget(
        'generate-kustomize-override',
        new GenerateKustomizeOverride(
            'k8s/base/deployment.yaml',
            'k8s/overlays/dev',
            [
                new KustomizeOverride('$.spec.template.metadata.labels.app', fn() => 'svc-events'),
                new KustomizeOverride(
                    '$.spec.template.spec.initContainers[0]',
                    fn(Context $c) => [
                        'name' => 'migrations',
                        'image' => $c->artifactCollector()->getByKey(__DIR__, 'image-migrations')->name()
                    ]
                ),
                new KustomizeOverride(
                    '$.spec.template.spec.containers[0]',
                    fn(Context $c) => [
                        'name' => 'app',
                        'image' => $c->artifactCollector()->getByKey(__DIR__, 'image-service')->name()
                    ]
                ),
            ]
        ),
        [
            new TargetId(__DIR__, 'image-service-docker-build'),
            new TargetId(__DIR__, 'image-migrations-docker-build'),
        ],
    );

    $builder->addTarget(
        'deploy',
        new KustomizeApply('k8s/overlays/dev'),
        [new TargetId(__DIR__, 'build')]
    );

    $builder->addDefaultTarget(
        DefaultTargetKind::Build,
        [
            new TargetId(__DIR__, 'generate-kustomize-override'),
        ]
    );

    $builder->addDefaultTarget(
        DefaultTargetKind::Fix
    );
};
