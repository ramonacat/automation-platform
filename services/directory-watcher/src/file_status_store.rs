use platform::mounts::PathInside;
use std::ops::Sub;
use std::sync::Arc;
use time::OffsetDateTime;
use tokio::sync::Mutex;
use tokio_postgres::Client;

#[derive(Debug, Copy, Clone)]
pub enum FileStatusSyncResult {
    Created,
    Modified,
    NotModified,
}

#[derive(Error, Debug)]
pub enum Error {
    #[error("Problems communicating with the database")]
    Database(#[from] tokio_postgres::Error),
}

#[async_trait]
pub trait FileStatusStore {
    async fn delete(&mut self, path: &PathInside) -> Result<(), Error>;
    async fn rename(&mut self, from: &PathInside, to: &PathInside) -> Result<(), Error>;
    async fn sync(
        &mut self,
        path: &PathInside,
        modified_at: OffsetDateTime,
    ) -> Result<FileStatusSyncResult, Error>;
}

pub struct Postgres {
    pg_client: Arc<Mutex<Client>>,
}
impl Postgres {
    pub fn new(pg_client: Arc<Mutex<Client>>) -> Self {
        Self { pg_client }
    }
}

#[async_trait]
impl FileStatusStore for Postgres {
    async fn delete(&mut self, path: &PathInside) -> Result<(), Error> {
        self.pg_client
            .lock()
            .await
            .execute(
                "DELETE FROM files WHERE mount_id=$1 AND path=$2",
                &[&path.mount_id(), &path.path()],
            )
            .await?;
        Ok(())
    }

    async fn rename(&mut self, from: &PathInside, to: &PathInside) -> Result<(), Error> {
        assert_eq!(
            from.mount_id(),
            to.mount_id(),
            "File moved between different mounts"
        );

        self.pg_client
            .lock()
            .await
            .execute(
                "UPDATE files SET path=$1 WHERE mount_id=$2 AND path=$3",
                &[&to.path(), &from.mount_id(), &from.path()],
            )
            .await?;
        Ok(())
    }

    async fn sync(
        &mut self,
        path: &PathInside,
        modified_at: OffsetDateTime,
    ) -> Result<FileStatusSyncResult, Error> {
        let mut postgres = self.pg_client.lock().await;
        let transaction = postgres.transaction().await?;

        let rows = transaction
            .query(
                "SELECT modified_date FROM files WHERE mount_id=$1 AND path=$2 FOR UPDATE",
                &[&path.mount_id(), &path.path()],
            )
            .await?;

        if let Some(row) = rows.get(0) {
            let current_modified_at = row.get::<_, OffsetDateTime>(0);
            if current_modified_at.sub(modified_at).abs() > std::time::Duration::from_secs(5) {
                transaction
                    .execute(
                        "UPDATE files SET modified_date=$1 WHERE mount_id=$2 AND path=$3",
                        &[&modified_at, &path.mount_id(), &path.path()],
                    )
                    .await?;

                transaction.commit().await?;
                return Ok(FileStatusSyncResult::Modified);
            }

            transaction.commit().await?;
            Ok(FileStatusSyncResult::NotModified)
        } else {
            transaction
                .execute("INSERT INTO files (id, mount_id, path, modified_date) VALUES(gen_random_uuid(), $1, $2, $3)", &[&path.mount_id(), &path.path(), &modified_at])
                .await?;

            transaction.commit().await?;
            Ok(FileStatusSyncResult::Created)
        }
    }
}
