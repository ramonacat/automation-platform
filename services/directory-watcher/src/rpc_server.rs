use std::{
    ops::Add,
    time::{Duration, SystemTime},
};

use events::{Client, Rpc as EventsRpc};
use lib_directory_watcher::{FilesystemEvent, FilesystemEventKind, Metadata, Rpc};
use platform::mounts::PathInside;
use rpc_support::{rpc_error::RpcError, RawRpcClient};
use time::OffsetDateTime;
use uuid::Uuid;

use crate::file_status_store::FileStatusStore;

pub struct RpcServer<T: FileStatusStore + Sync + Send, TRawRpcClient: RawRpcClient + Send + Sync> {
    file_status_store: T,
    event_service: Client<TRawRpcClient>,
    generate_uuid: Box<dyn Fn() -> Uuid + Send + Sync>,
}

impl<T: FileStatusStore + Sync + Send, TRawRpcClient: RawRpcClient + Send + Sync>
    RpcServer<T, TRawRpcClient>
{
    pub const fn new(
        file_status_store: T,
        event_service: events::Client<TRawRpcClient>,
        generate_uuid: Box<dyn Fn() -> Uuid + Send + Sync>,
    ) -> Self {
        Self {
            file_status_store,
            event_service,
            generate_uuid,
        }
    }
}

#[async_trait::async_trait]
impl<T: FileStatusStore + Sync + Send, TRawRpcClient: RawRpcClient + Send + Sync> Rpc
    for RpcServer<T, TRawRpcClient>
{
    #[allow(clippy::too_many_lines)]
    async fn file_changed(
        &mut self,
        event: FilesystemEvent,
        _metadata: Metadata,
    ) -> Result<(), RpcError> {
        info!("Received file changed event: {:?}", event);

        match event.kind {
            FilesystemEventKind::Created {} | FilesystemEventKind::Modified {} => {
                let timestamp = OffsetDateTime::from_unix_timestamp(
                    event
                        .timestamp
                        .try_into()
                        .map_err(|e| RpcError::Custom(format!("{e}")))?,
                )
                .map_err(|e| RpcError::Custom(format!("{e}")))?;

                let sync_status = self
                    .file_status_store
                    .sync(
                        &PathInside::new(event.mount_id.clone(), event.path.clone()),
                        timestamp,
                    )
                    .await
                    .map_err(|e| RpcError::Custom(format!("{e}")))?;

                let data = match sync_status {
                    crate::file_status_store::FileStatusSyncResult::Created => {
                        events::EventKind::FileCreated {
                            path: events::FileOnMountPath {
                                path: event.path,
                                mount_id: event.mount_id,
                            },
                        }
                    }
                    crate::file_status_store::FileStatusSyncResult::Modified
                    | crate::file_status_store::FileStatusSyncResult::NotModified => {
                        events::EventKind::FileChanged {
                            path: events::FileOnMountPath {
                                path: event.path,
                                mount_id: event.mount_id,
                            },
                        }
                    }
                };

                self.event_service
                    .send_event(
                        events::Event {
                            id: (self.generate_uuid)(),
                            created_time: timestamp.into(),
                            data,
                        },
                        events::Metadata {
                            // TODO pass the correlation ID from parent scope!
                            correlation_id: (self.generate_uuid)(),
                            source: "directory-watcher".to_string(),
                        },
                    )
                    .await?;
            }
            FilesystemEventKind::Moved { to } => {
                self.file_status_store
                    .rename(
                        &PathInside::new(event.mount_id.clone(), event.path.clone()),
                        &PathInside::new(event.mount_id.clone(), to.clone()),
                    )
                    .await
                    .map_err(|e| RpcError::Custom(format!("{e}")))?;

                self.event_service
                    .send_event(
                        events::Event {
                            id: (self.generate_uuid)(),
                            created_time: SystemTime::UNIX_EPOCH
                                .add(Duration::from_secs(event.timestamp)),
                            data: events::EventKind::FileMoved {
                                from: events::FileOnMountPath {
                                    path: event.path,
                                    mount_id: event.mount_id.clone(),
                                },
                                to: events::FileOnMountPath {
                                    path: to,
                                    mount_id: event.mount_id,
                                },
                            },
                        },
                        events::Metadata {
                            // TODO pass the correlation ID from parent scope!
                            correlation_id: (self.generate_uuid)(),
                            source: "directory-watcher".to_string(),
                        },
                    )
                    .await?;
            }
            FilesystemEventKind::Deleted {} => {
                self.file_status_store
                    .delete(&PathInside::new(event.mount_id.clone(), event.path.clone()))
                    .await
                    .map_err(|e| RpcError::Custom(format!("{e}")))?;

                self.event_service
                    .send_event(
                        events::Event {
                            id: (self.generate_uuid)(),
                            created_time: SystemTime::UNIX_EPOCH
                                .add(Duration::from_secs(event.timestamp)),
                            data: events::EventKind::FileDeleted {
                                path: events::FileOnMountPath {
                                    path: event.path,
                                    mount_id: event.mount_id,
                                },
                            },
                        },
                        events::Metadata {
                            // TODO pass the correlation ID from parent scope!
                            correlation_id: (self.generate_uuid)(),
                            source: "directory-watcher".to_string(),
                        },
                    )
                    .await?;
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;
    use std::sync::Arc;

    use rpc_support::ResponseStream;
    use serde::{de::DeserializeOwned, Serialize};
    use tokio::sync::Mutex;

    use super::*;
    use crate::file_status_store::Error;
    use crate::file_status_store::FileStatusSyncResult;

    struct MockFileStatusStore {
        pub events: Vec<String>,
        pub sync_result: FileStatusSyncResult,
    }

    impl MockFileStatusStore {
        pub fn new(sync_result: FileStatusSyncResult) -> Self {
            Self {
                events: vec![],
                sync_result,
            }
        }
    }

    #[async_trait::async_trait]
    impl FileStatusStore for MockFileStatusStore {
        async fn delete(&mut self, path: &PathInside) -> Result<(), Error> {
            self.events.push(format!("delete {path:?}"));

            Ok(())
        }

        async fn rename(&mut self, from: &PathInside, to: &PathInside) -> Result<(), Error> {
            self.events.push(format!("rename {from:?} to {to:?}"));

            Ok(())
        }

        async fn sync(
            &mut self,
            path: &PathInside,
            modified_at: OffsetDateTime,
        ) -> Result<FileStatusSyncResult, Error> {
            self.events
                .push(format!("sync {path:?} at {modified_at:?}"));

            Ok(self.sync_result)
        }
    }

    struct MockRawRpcClient {
        pub rpcs: Arc<Mutex<Vec<String>>>,
    }

    #[async_trait::async_trait]
    impl RawRpcClient for MockRawRpcClient {
        // TODO rename to send_rpc_request
        async fn send_rpc<TRequest, TMetadata, TResponse>(
            &mut self,
            id: u64,
            method_name: &str,
            request: &TRequest,
            metadata: &TMetadata,
        ) -> Result<TResponse, RpcError>
        where
            TRequest: Serialize + Sync + Send,
            TMetadata: Serialize + Sync + Send,
            TResponse: DeserializeOwned,
        {
            self.rpcs.lock().await.push(format!(
                "send_rpc {id} {method_name} {:?} {:?}",
                serde_json::to_value(request),
                serde_json::to_value(metadata)
            ));

            Ok(serde_json::from_str("null").unwrap())
        }

        async fn send_rpc_stream_request<TRequest, TMetadata, TResponse>(
            &mut self,
            _request_id: u64,
            _method_name: &str,
            _request: &TRequest,
            _metadata: &TMetadata,
        ) -> Result<ResponseStream<TResponse>, RpcError>
        where
            TRequest: Serialize + Sync + Send,
            TMetadata: Serialize + Sync + Send,
            TResponse: DeserializeOwned,
        {
            todo!()
        }
    }

    #[tokio::test]
    async fn test_file_created() {
        let file_status_store = MockFileStatusStore::new(FileStatusSyncResult::Created);
        let rpcs = Arc::new(Mutex::new(vec![]));
        let raw_rpc_client = MockRawRpcClient { rpcs: rpcs.clone() };

        let event_service = events::Client::new(raw_rpc_client);

        let mut rpc_server = RpcServer::new(
            file_status_store,
            event_service,
            Box::new(|| Uuid::from_str("00000000-0000-0000-0000-000000000000").unwrap()),
        );

        let event = FilesystemEvent {
            kind: FilesystemEventKind::Created {},
            path: "/test".to_string(),
            mount_id: "test".to_string(),
            timestamp: 1024,
        };

        rpc_server.file_changed(event, Metadata {}).await.unwrap();

        insta::assert_debug_snapshot!(rpc_server.file_status_store.events);

        let rpcs = rpcs.lock().await;

        insta::assert_debug_snapshot!(rpcs.clone());
    }

    #[tokio::test]
    async fn test_file_moved() {
        let file_status_store = MockFileStatusStore::new(FileStatusSyncResult::Created);
        let rpcs = Arc::new(Mutex::new(vec![]));
        let raw_rpc_client = MockRawRpcClient { rpcs: rpcs.clone() };

        let event_service = events::Client::new(raw_rpc_client);

        let mut rpc_server = RpcServer::new(
            file_status_store,
            event_service,
            Box::new(|| Uuid::from_str("00000000-0000-0000-0000-000000000000").unwrap()),
        );

        let event = FilesystemEvent {
            kind: FilesystemEventKind::Moved {
                to: "/test2".to_string(),
            },
            path: "/test".to_string(),
            mount_id: "test".to_string(),
            timestamp: 1024,
        };

        rpc_server.file_changed(event, Metadata {}).await.unwrap();

        insta::assert_debug_snapshot!(rpc_server.file_status_store.events);

        let rpcs = rpcs.lock().await;

        insta::assert_debug_snapshot!(rpcs.clone());
    }

    #[tokio::test]
    async fn test_file_deleted() {
        let file_status_store = MockFileStatusStore::new(FileStatusSyncResult::Created);
        let rpcs = Arc::new(Mutex::new(vec![]));
        let raw_rpc_client = MockRawRpcClient { rpcs: rpcs.clone() };

        let event_service = events::Client::new(raw_rpc_client);

        let mut rpc_server = RpcServer::new(
            file_status_store,
            event_service,
            Box::new(|| Uuid::from_str("00000000-0000-0000-0000-000000000000").unwrap()),
        );

        let event = FilesystemEvent {
            kind: FilesystemEventKind::Deleted {},
            path: "/test".to_string(),
            mount_id: "test".to_string(),
            timestamp: 1024,
        };

        rpc_server.file_changed(event, Metadata {}).await.unwrap();

        insta::assert_debug_snapshot!(rpc_server.file_status_store.events);

        let rpcs = rpcs.lock().await;

        insta::assert_debug_snapshot!(rpcs.clone());
    }

    #[tokio::test]
    async fn test_file_modified() {
        let file_status_store = MockFileStatusStore::new(FileStatusSyncResult::Modified);
        let rpcs = Arc::new(Mutex::new(vec![]));
        let raw_rpc_client = MockRawRpcClient { rpcs: rpcs.clone() };

        let event_service = events::Client::new(raw_rpc_client);

        let mut rpc_server = RpcServer::new(
            file_status_store,
            event_service,
            Box::new(|| Uuid::from_str("00000000-0000-0000-0000-000000000000").unwrap()),
        );

        let event = FilesystemEvent {
            kind: FilesystemEventKind::Modified {},
            path: "/test".to_string(),
            mount_id: "test".to_string(),
            timestamp: 1024,
        };

        rpc_server.file_changed(event, Metadata {}).await.unwrap();

        insta::assert_debug_snapshot!(rpc_server.file_status_store.events);

        let rpcs = rpcs.lock().await;

        insta::assert_debug_snapshot!(rpcs.clone());
    }
}
