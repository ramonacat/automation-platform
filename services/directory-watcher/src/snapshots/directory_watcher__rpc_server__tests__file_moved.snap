---
source: src/rpc_server.rs
expression: rpc_server.file_status_store.events
---
[
    "rename PathInside { mount_id: \"test\", path: \"/test\" } to PathInside { mount_id: \"test\", path: \"/test2\" }",
]
